### DISCUSSION TITLE
Compounding with CJB vs Reinvesting with 3Commas

### DISCUSSION THREAD

**Discussion by MisterMelons** [cyberjunky/3commas-cyber-bots#209] [General]
Hello,

I'm having an issue using the Reinvesting feature on 3Commas and their support is having difficulty finding the culprit. Please bear in mind, I could be completely wrong about what I'm about to say, but it's just my observation...

So far, I prefer to use 3Comma's Reinvesting feature because they split the profit of a bot evenly over the Base Order and all Safeties configured. This is quite different than using CJB's Martingale for compounding where smaller amounts are reinvested in the beginning of a new deal.  My issue is that my 3C bots rarely Reinvest for an unknown reason. Here is the difference in the two solutions.

Reinvesting with 3Commas:

If I make $15 in profits, 3C will respect my original Base Order and Safety Order amounts, but then they will split the $15 and add that amount to the BO and SO orders when a new deal is made. If I had a BO and 3 SO's they would add $3.75 to my BO, and $3.75 to each SO. Compared to Martingale I prefer this method because it gives more capital to the early buys, those being BO and SO1.

Compounding with CJB:

If I make $15 in profits, compound.py will rewrite my BO and SO amounts to include the Martingale step process. With the same BO and 3 SO's above, my BO would increase by $1, SO1 by $2, SO2 by $4 and SO3 by $8. My problem with this is that I rarely get into SO3, where most of the compounded capital would go towards. Using Martingale my compounded capital doesn't get reinvested as much as 3Comma's method, i.e. my BO only increases by $1 instead of $3.75 and SO1 only increases by $2 instead of an additional $3.75. I'd like to have that compounded capital working more towards the front of the deal.

I was using CJB Compounding, but when I went back to 3C Reinvesting I noticed that 3C rarely reinvests anything. I've been talking with 3C support for the last week and they can't tell me for certain why this is happening. At one point I was led to believe that stopping and starting a bot would reset the profit tracking for reinvestment. This made me wonder if using botupdater.py to update my bot pairs every hour was the culprit. That doesn't seem to be the case though because this is what support just told me:

```
As it turns out:

1. If you turn off a DCA bot and re-enable it again, reinvestment continues from where it previously left of.
2. If you just edit the DCA bot (just added/removed pairs), bot is enabled and reinvestment still On, 
the reinvestment will still continue as it was

The reinvestment will only reset if the profit currency changed.
```

I'm still waiting for the higher tier 3C developers to get back to me.

Regardless, my question is:

Can CJB's **compound.py** employ an alternative method for compounding besides using Martingale? I see that martingale is mentioned in the code, as if it were calling a function that does the math for martingale. If that's the case, is there a way to change to 3C's style of Reinvesting? I'm not sure if this is possible with scripts outside of 3Commas because as I said, 3C is tracking the profit split and adding it on top of the default BO and SO amounts _at deal creation time_. I don't know if CJB could have the same ability.

I trust CJB's ability to reinvest more than I can trust 3Commas, so the above option would be fantastic!

Anyway, thanks for any help. I hope what I wrote makes sense!

**Comment by MisterMelons** [comment]
After weeks of troubleshooting this issue with 3Commas, it appears that 3C will **NOT** Reinvest if a deal is closed as a Stop Loss. When a deal is closed by Stop Loss, the only 3C feature that works is **Risk Reduction**.

I suggested that 3Commas use a conditional check when a deal is closed using Stop Loss. If the deal closes with a profit, it should carry on with **Reinvestment**, but if the Stop Loss resulted in a loss in profit then it should carry on with Risk Reduction.

The alternative I would prefer is using CJB with non-Martingale as described above...

**Comment by MisterMelons** [comment]
UPDATE: 3Commas devs got back to me this morning. The issue with not Reinvesting has nothing to do with deals closed by Stop Loss. It appears that the **Reinvestment tracking is reset every time you update the pairs in a bot**. Since I do that every 15 minutes my only option is CJB Compounding.

Just talking to myself... and the guy that comes here in 2 years.

**Comment by amargedon** [comment]
Not that active over here as my focus shifted to Freqtrade, and no longer so much 3C.

I think the compound script can be updated with an alternate way of applying compounding of profit (or loss). However I've a serious concern when the bot is used with a martingale scale, and the profit is not compounded that way. This would change the required change to close the deal in unexpected ways because the bo:so is changed. 

Why do you not configure another martingale factor in the bot, as that seems what you want to do anyway? In that way the compound script can work like it is already for you?
